### PR TITLE
1 create load balancer terraform module

### LINKED ISSUE #1: Create Load Balancer Terraform module

### PR DESCRIPTION
**Pull request by Browix**
## What changes does your PR bring?

In a few words, describe what changes your commits make to the code.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (non-breaking change to the build process or auxiliary tools)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation (Improvements or additions to documentation)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## References

Fixes:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1 +1,27 @@
...
+  region      = var.region
+  size        = var.size
+  size_unit   = var.size_unit
+  vpc_uuid    = var.vpc_uuid
+  droplet_ids = var.droplet_ids
+  droplet_tag = var.droplet_tag
+
+  dynamic "forwarding_rule" {
+    for_each = var.forwarding_rule
+    content {
```
Comment (stawii) [inline main.tf:23]: The `tls_passthrough` is missing but seems usable if one has TLS encryption on host level ie. ingress-controller or any ssl-cert on application level.

File: main.tf [hunk 2]
```diff
@@ -1 +1,27 @@
-# Terraform code goes here
+resource "digitalocean_loadbalancer" "main" {
```
Comment (stawii) [inline main.tf:12]: Few options are missing ie: `algorithm` and `redirect_http_to_https` that sound useful. But we can add them when actually needed, not to bloat too much now.

File: variables.tf [hunk 3]
```diff
@@ -1,12 +1,51 @@
...
+}
+
+variable "droplet_tag" {
+  description = "The name of a Droplet tag corresponding to Droplets to be assigned to the Load Balancer."
+  type        = string
+  default     = null
+}
+
+variable "firewall" {
+  description = "List of objects that represent the configuration of each healthcheck."
```
Comment (stawii) [inline variables.tf]: Description is missleading, its about firewall not healthchecks. Aaaand healthchecks are missing ;)

File: variables.tf [hunk 4]
```diff
@@ -1,12 +1,51 @@
...
+variable "size_unit" {
+  description = "The size of the Load Balancer. It must be in the range (1, 100). Defaults to 1. Only one of size or size_unit may be provided."
+  type        = number
+  default     = 1
+}
+
+variable "vpc_uuid" {
+  description = "The ID of the VPC where the load balancer will be located."
+  type        = string
+  default     = ""
```
Comment (stawii) [inline variables.tf]: Either null, as "real" empty value, or no default at all. Lets set it to null, so it's the same as in droplets and k8s.

File: main.tf [hunk 5]
```diff
@@ -1 +1,27 @@
...
+  }
+
+  dynamic "firewall" {
+    for_each = var.firewall
+    content {
+      deny  = lookup(firewall.value, "deny", null)
+      allow = lookup(firewall.value, "allow", null)
+    }
+  }
+}
```
Comment (stawii) [inline main.tf:50]: The `healthcheck` block is missing.

File: variables.tf [hunk 6]
```diff
@@ -1,12 +1,51 @@
...
+
+variable "droplet_tag" {
+  description = "The name of a Droplet tag corresponding to Droplets to be assigned to the Load Balancer."
+  type        = string
+  default     = null
+}
+
+variable "firewall" {
+  description = "List of objects that represent the configuration of each healthcheck."
+  type        = list(any)
```
Comment (stawii) [inline variables.tf:48]: Please break it down to two vars: `firewall_deny_rules` and `firewall_allow_rules`. Then both would be simple sets/lists rather than objects - with undefined structure.

File: versions.tf
Comment (stawii) [inline versions.tf:1]: Move this content to begining of main.tf

File: variables.tf [hunk 7]
```diff
@@ -1,12 +1,51 @@
...
-# }
-
-# variable "variable_password" {
-#	  description = "variable description"
-#   type        = string
-#   sensitive   = true
-#   default     = "abc"
-# }
+variable "name" {
+  description = "The Load Balancer name"
```
Comment (stawii) [inline variables.tf]: Missing dot.

File: variables.tf [hunk 8]
```diff
@@ -1,12 +1,51 @@
...
+  type        = string
+}
+
+variable "region" {
+  description = "The region to start in"
```
Comment (stawii) [inline variables.tf]: Missing dot.

File: examples/basic/main.tf [hunk 9]
```diff
@@ -0,0 +1,39 @@
+provider "digitalocean" {
+  token = var.do_token
+}
+
+module "terraform_module_digitalocean_vpc" {
```
Comment (stawii) [inline examples/basic/main.tf]: Lets shorten this to `digitalocean_vpc` or even something like `do_vpc` or `example_vpc`.

File: examples/basic/main.tf [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  source      = "github.com/opsd-io/terraform-module-digitalocean-vpc"
+  vpc_name    = "your-vpc"
+  region      = "nyc1"
+  ip_range    = "192.168.0.0/24"
+  description = "VPC added by terraform module"
+}
+
+
+module "digitalocean_droplet" {
+  source   = "github.com/opsd-io/terraform-module-digitalocean-droplet?ref=1-create-droplet-terraform-module"
```
Comment (stawii) [inline examples/basic/main.tf]: Don't pin to feature branch.

File: examples/basic/main.tf [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  image    = "ubuntu-20-04-x64"
+  name     = "web-1"
+  region   = "nyc1"
+  size     = "s-1vcpu-1gb"
+  vpc_uuid = module.terraform_module_digitalocean_vpc.id
+}
+
+module "digitalocean_loadbalancer" {
+  source      = "github.com/opsd-io/terraform-module-digitalocean-load-balancer?ref=1-create-load-balancer-terraform-module"
```
Comment (stawii) [inline examples/basic/main.tf]: Don't pin to feature branch.

File: variables.tf [hunk 12]
```diff
@@ -1,12 +1,51 @@
...
+  default     = 1
+}
+
+variable "vpc_uuid" {
+  description = "The ID of the VPC where the load balancer will be located."
+  type        = string
+  default     = ""
+}
+
+variable "forwarding_rule" {
```
Comment (stawii) [inline variables.tf]: it is list, so plural, change to `forwarding_rules`.

File: .tool-versions [hunk 13]
```diff
@@ -2,3 +2,4 @@ terraform 1.5.7
 terraform-docs 0.17.0
 tflint 0.50.3
 pre-commit 3.6.2
+python 3.12.1
```
Comment (stawii) [inline .tool-versions]: Remove python from here, we don't need it.

File: examples/basic/variables.tf [hunk 14]
```diff
@@ -0,0 +1,10 @@
+variable "do_token" {
```
Comment (stawii) [inline examples/basic/variables.tf:1]: It's fine, but please add `default = null` so provider fallbacks to searching env variables.

File: examples/basic/variables.tf [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+  description = "Token to digitalOcean API"
+  type        = string
+}
+
+variable "droplet_ids" {
```
Comment (stawii) [inline examples/basic/variables.tf]: It's not used in example - there is just one static droplet defined, please remove this var.

**Review by stawii — CHANGES_REQUESTED** [review]

File: examples/basic/variables.tf (hunk 15 above)
Comment (Browix) [reply examples/basic/variables.tf]: removed

File: examples/basic/variables.tf (hunk 14 above)
Comment (Browix) [reply examples/basic/variables.tf:1]: added

File: .tool-versions (hunk 13 above)
Comment (Browix) [reply .tool-versions]: ok

File: variables.tf (hunk 8 above)
Comment (Browix) [reply variables.tf]: corrected

File: examples/basic/main.tf (hunk 9 above)
Comment (Browix) [reply examples/basic/main.tf]: corrected

File: examples/basic/main.tf (hunk 10 above)
Comment (Browix) [reply examples/basic/main.tf]: corrected

File: variables.tf (hunk 12 above)
Comment (Browix) [reply variables.tf]: changed

File: variables.tf (hunk 4 above)
Comment (Browix) [reply variables.tf]: ok

File: main.tf (hunk 5 above)
Comment (Browix) [reply main.tf:50]: added health-check

File: variables.tf (hunk 6 above)
Comment (Browix) [reply variables.tf:48]: changed :
firewall {
    allow = var.firewall_allow
    deny  = var.firewall_deny
  }

File: variables.tf (hunk 3 above)
Comment (Browix) [reply variables.tf]: corrected

File: main.tf (hunk 1 above)
Comment (Browix) [reply main.tf:23]: added

**Review by stawii — APPROVED** [review]
